src/function/gds/louvain.cpp

```diff
@@ -0,0 +1,605 @@
+#include <cmath>
+
+#include "binder/binder.h"
+#include "common/exception/runtime.h"
+#include "function/gds/gds_function_collection.h"
+#include "function/gds/gds_utils.h"
+#include "function/gds/gds_utils_mem.h"
+#include "gds_vertex_compute.h"
+#include "graph/graph_mem.h"
+#include "processor/execution_context.h"
+
+using namespace std;
+using namespace kuzu::binder;
+using namespace kuzu::common;
+using namespace kuzu::processor;
+using namespace kuzu::storage;
+using namespace kuzu::graph;
+
+// Louvain method for community detection: https://en.wikipedia.org/wiki/Louvain_method.
+// For nodes in a graph grouped into `C` communities, the modularity of a graph is given as:
+//   modularity = sumIntraWeights/2m - (sumWeightedDegrees/2m)^2
+// where:
+//   sumIntraWeights = sum over all c in C: sum of all edges in c, counted twice in both directions
+//                   = sum of the weightedDegree of all nodes considering only edges to neighbors
+//                     that are in the same community
+//   sumWeightedDegrees = sum over all c in C: (sum of weightedDegree of all nodes in c)^2
+//   weightedDegree(n) = sum of the weight of all edges attached to node n
+
+// The parallel Louvain implementation here follows https://hpc.pnl.gov/people/hala/grappolo.html.
+
+namespace kuzu {
+namespace function {
+
+constexpr uint32_t MAX_PHASES = 10;
+constexpr uint32_t MAX_ITERATIONS = 10;
+constexpr double THRESHOLD = 1e-6;
+static constexpr offset_t UNASSIGNED_COMM = numeric_limits<offset_t>::max();
+
+class CommInfo {
+public:
+    CommInfo() : size{0}, degree(0) {}
+    CommInfo(const CommInfo& other) {
+        size.store(other.getSize());
+        degree.store(other.getDegree());
+    }
+    CommInfo& operator=(const CommInfo& other) {
+        if (this != &other) {
+            size.store(other.getSize());
+            degree.store(other.getDegree());
+        }
+        return *this;
+    }
+
+    offset_t getSize() const { return size.load(memory_order_relaxed); }
```

Comment:
Why is this memory_order_relaxed? I thought `size` can be changing as another thread is updating the size. That looks like a clear case of seq_order load.

Reply:
The code does not read and write concurrently. The order is fetch_add in parallel -> read in parallel.